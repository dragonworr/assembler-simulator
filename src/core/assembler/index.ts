import { produce } from 'immer'
import { tokenize } from './tokenizer'
import type { Statement } from './parser'
import { OperandType, parse } from './parser'
import {
  DuplicateLabelError,
  EndOfMemoryError,
  LabelNotExistError,
  JumpDistanceError
} from '../exceptions'
import { Mnemonic } from '../constants'
import { exp } from '../../common/utils'

export * from './tokenizer'
export * from './parser'

type LabelToAddressMap = Record<string, number>

const getLabelToAddressMap = (statements: Statement[]): LabelToAddressMap => {
  const [, labelToAddressMap] = statements.reduce<[number, LabelToAddressMap]>(
    ([address, labelToAddressMap], statement, index) => {
      const { label, instruction, operands, machineCodes } = statement
      const firstOperand = operands[0]
      return [
        instruction.token.value === Mnemonic.ORG
          ? (firstOperand.value as number)
          : exp<number>(() => {
              const nextAddress =
                address +
                machineCodes.length +
                (firstOperand !== undefined && firstOperand.type === OperandType.Label ? 1 : 0)
              if (nextAddress > 0xff && index !== statements.length - 1) {
                throw new EndOfMemoryError(statement)
              }
              return nextAddress
            }),
        label === null
          ? labelToAddressMap
          : produce(labelToAddressMap, draft => {
              if (draft[label.identifier] !== undefined) {
                throw new DuplicateLabelError(label)
              }
              draft[label.identifier] = address
            })
      ]
    },
    [0, {}]
  )
  return labelToAddressMap
}

export type AddressToMachineCodeMap = Record<number, number>
export type AddressToStatementMap = Record<number, Statement>

type AssembleResult = [AddressToMachineCodeMap, AddressToStatementMap]

export const assemble = (input: string): AssembleResult => {
  const statements = parse(tokenize(input))
  const labelToAddressMap = getLabelToAddressMap(statements)
  const [, addressToMachineCodeMap, addressToStatementMap] = statements.reduce<
    [number, ...AssembleResult]
  >(
    ([address, addressToMachineCodeMap, addressToStatementMap], statement) => {
      const { instruction, operands, machineCodes } = statement
      const firstOperand = operands[0]
      if (instruction.token.value === Mnemonic.ORG) {
        return [firstOperand.value as number, addressToMachineCodeMap, addressToStatementMap]
      }
      if (firstOperand !== undefined && firstOperand.type === OperandType.Label) {
        const labelAddress = labelToAddressMap[firstOperand.token.value]
        if (labelAddress === undefined) {
          throw new LabelNotExistError(firstOperand.token)
        }
        const distance = labelAddress - address
        if (distance < -128 || distance > 127) {
          throw new JumpDistanceError(firstOperand.token)
        }
        const unsignedDistance = distance < 0 ? 0x100 + distance : distance
        machineCodes.push(unsignedDistance)
      }
      return [
        address + machineCodes.length,
        produce(addressToMachineCodeMap, draft => {
          machineCodes.forEach((machineCode, index) => {
            draft[address + index] = machineCode
          })
        }),
        produce(addressToStatementMap, draft => {
          draft[address] = statement
        })
      ]
    },
    [0, {}, {}]
  )
  return [addressToMachineCodeMap, addressToStatementMap]
}
