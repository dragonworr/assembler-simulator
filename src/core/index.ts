export { assemble } from './assembler'
export * from './memory'
export * from './cpu'
