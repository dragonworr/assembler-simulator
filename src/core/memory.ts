import type { AddressToMachineCodeMap } from './assembler'

export const initMemory = (): number[] =>
  [...Array(0x100)].map((_, address) => (address < 0xc0 ? 0 : 0x20))

export const initMemoryFrom = (map: AddressToMachineCodeMap): number[] => {
  const memory = initMemory()
  map.forEach((machineCode, address) => {
    memory[address] = machineCode
  })
  return memory
}
