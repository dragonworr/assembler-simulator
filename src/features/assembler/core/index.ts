import { produce } from 'immer'
import { tokenize } from './tokenizer'
import { OperandType, Statement, parse } from './parser'
import {
  DuplicateLabelError,
  AssembleEndOfMemoryError,
  LabelNotExistError,
  JumpDistanceError
} from '../../../common/exceptions'
import { Mnemonic } from '../../../common/constants'
import { call } from '../../../common/utils'

export * from './types'
export * from './tokenizer'
export * from './parser'

type LabelToAddressMap = Record<string, number>

const getLabelToAddressMap = (statements: Statement[]): LabelToAddressMap => {
  const [, labelToAddressMap] = statements.reduce<[number, LabelToAddressMap]>(
    ([address, labelToAddressMap], statement, index) => {
      const { label, instruction, operands, machineCodes } = statement
      const firstOperand = operands[0]
      return [
        instruction.mnemonic === Mnemonic.ORG
          ? (firstOperand.value as number)
          : call((): number => {
              const nextAddress =
                address +
                machineCodes.length +
                (firstOperand !== undefined && firstOperand.type === OperandType.Label ? 1 : 0)
              if (nextAddress > 0xff && index !== statements.length - 1) {
                throw new AssembleEndOfMemoryError(statement)
              }
              return nextAddress
            }),
        label === null
          ? labelToAddressMap
          : produce(labelToAddressMap, draft => {
              if (draft[label.identifier] !== undefined) {
                throw new DuplicateLabelError(label)
              }
              draft[label.identifier] = address
            })
      ]
    },
    [0, {}]
  )
  return labelToAddressMap
}

export type AddressToMachineCodeMap = Record<number, number>
export type AddressToStatementMap = Record<number, Statement>

type AssembleResult = [AddressToMachineCodeMap, AddressToStatementMap]

export const assemble = (input: string): AssembleResult => {
  const statements = parse(tokenize(input))
  const labelToAddressMap = getLabelToAddressMap(statements)
  const [, addressToMachineCodeMap, addressToStatementMap] = statements.reduce<
    [number, ...AssembleResult]
  >(
    ([address, addressToMachineCodeMap, addressToStatementMap], statement) => {
      const { instruction, operands, machineCodes } = statement
      const firstOperand = operands[0]
      if (instruction.mnemonic === Mnemonic.ORG) {
        return [firstOperand.value as number, addressToMachineCodeMap, addressToStatementMap]
      }
      if (firstOperand !== undefined && firstOperand.type === OperandType.Label) {
        const labelAddress = labelToAddressMap[firstOperand.rawValue]
        if (labelAddress === undefined) {
          throw new LabelNotExistError(firstOperand)
        }
        const distance = labelAddress - address
        if (distance < -128 || distance > 127) {
          throw new JumpDistanceError(firstOperand)
        }
        const unsignedDistance = distance < 0 ? 0x100 + distance : distance
        machineCodes.push(unsignedDistance)
      }
      return [
        address + machineCodes.length,
        produce(addressToMachineCodeMap, draft => {
          machineCodes.forEach((machineCode, index) => {
            draft[address + index] = machineCode
          })
        }),
        produce(addressToStatementMap, draft => {
          draft[address] = statement
        })
      ]
    },
    [0, {}, {}]
  )
  return [addressToMachineCodeMap, addressToStatementMap]
}
