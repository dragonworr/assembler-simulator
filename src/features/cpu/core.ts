import { produce } from 'immer'
import {
  InvalidRegisterError,
  RunBeyondEndOfMemory,
  // StackOverflowError,
  // StackUnderflowError,
  DivideByZeroError
} from '../../common/exceptions'
import { Opcode, Register } from '../../common/constants'
import { Head, unsign8bit, exp } from '../../common/utils'

type GPR = [AL: number, BL: number, CL: number, DL: number]

const MAX_SP = 0xbf

enum Flag {
  Zero,
  Overflow,
  Sign,
  Interrupt
}

type SR = [zero: boolean, overflow: boolean, sign: boolean, interrupt: boolean]

interface CPU {
  gpr: GPR
  ip: number
  sp: number
  sr: SR
  isHalted: boolean
}

export const init = (): CPU => {
  return {
    gpr: [0, 0, 0, 0],
    ip: 0,
    sp: MAX_SP,
    sr: [false, false, false, false],
    isHalted: false
  }
}

const checkGPR = (register: number): Register => {
  if (register < 0 || register > 3) {
    throw new InvalidRegisterError(register)
  }
  return register
}

const checkIP = (address: number): number => {
  if (address > 0xff) {
    throw new RunBeyondEndOfMemory()
  }
  return address
}

// const checkSP = (address: number): number => {
//   if (address < 0) {
//     throw new StackOverflowError()
//   }
//   if (address > MAX_SP) {
//     throw new StackUnderflowError()
//   }
//   return address
// }

// const getFlagsValue = (sr: SR): number =>
//   sr.reduce((result, isSet, flag) => (isSet ? result + 0b10 ** flag : result), 0)

const checkOperationResult = (
  result: number,
  previousValue: number
): [result: number, flags: Partial<SR>] => {
  const finalResult = result > 0xff ? result % 0x100 : unsign8bit(result)
  const flags: Head<SR> = [false, false, false]
  if (finalResult === 0) {
    flags[Flag.Zero] = true
  } else if (finalResult >= 0x80) {
    flags[Flag.Sign] = true
  }
  if (
    /* addition */ (previousValue < 0x80 && result >= 0x80) ||
    /* subtraction */ (previousValue >= 0x80 && result < 0x80)
  ) {
    flags[Flag.Overflow] = true
  }
  return [finalResult, flags]
}

export const step = (__cpu: CPU, __memory: number[]): [cpu: CPU, memory: number[]] =>
  produce([__cpu, __memory], ([cpu, memory]) => {
    const getGPR = (register: Register): number => cpu.gpr[register]
    const setGPR = (register: Register, value: number): void => {
      cpu.gpr[register] = value
    }

    const setIP = (address: number): void => {
      cpu.ip = address
    }
    const incIP = (): void => {
      setIP(checkIP(cpu.ip + 1))
    }

    const setSR = (flags: Partial<SR>): void => {
      Object.assign(cpu.sr, flags)
    }

    const loadFromMemory = (address: number): number => {
      return memory[address]
    }
    // const storeToMemory = (address: number, machineCode: number): void => {
    //   memory[address] = machineCode
    // }

    /**
     * @modifies {@link cpu.ip}
     */
    const getNextOpcode = (): number => {
      incIP()
      return loadFromMemory(cpu.ip)
    }

    const opcode = loadFromMemory(cpu.ip)

    switch (opcode) {
      case Opcode.END:
        cpu.isHalted = true
        break
      case Opcode.ADD_REG_TO_REG: {
        const destReg = checkGPR(getNextOpcode())
        const srcReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(
          getGPR(destReg) + getGPR(srcReg),
          getGPR(destReg)
        )
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.SUB_REG_FROM_REG: {
        const destReg = checkGPR(getNextOpcode())
        const srcReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(
          getGPR(destReg) - getGPR(srcReg),
          getGPR(destReg)
        )
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.MUL_REG_BY_REG: {
        const destReg = checkGPR(getNextOpcode())
        const srcReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(
          getGPR(destReg) * getGPR(srcReg),
          getGPR(destReg)
        )
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.DIV_REG_BY_REG: {
        const destReg = checkGPR(getNextOpcode())
        const srcReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(
          exp<number>(() => {
            const divisor = getGPR(srcReg)
            if (divisor === 0) {
              throw new DivideByZeroError()
            }
            return Math.floor(getGPR(destReg) / divisor)
          }),
          getGPR(destReg)
        )
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.INC_REG: {
        const destReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(getGPR(destReg) + 1, getGPR(destReg))
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.DEC_REG: {
        const destReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(getGPR(destReg) - 1, getGPR(destReg))
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.MOD_REG_BY_REG: {
        const destReg = checkGPR(getNextOpcode())
        const srcReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(
          exp<number>(() => {
            const divisor = getGPR(srcReg)
            if (divisor === 0) {
              throw new DivideByZeroError()
            }
            return getGPR(destReg) % divisor
          }),
          getGPR(destReg)
        )
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.AND_REG_WITH_REG: {
        const destReg = checkGPR(getNextOpcode())
        const srcReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(
          getGPR(destReg) & getGPR(srcReg),
          getGPR(destReg)
        )
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.OR_REG_WITH_REG: {
        const destReg = checkGPR(getNextOpcode())
        const srcReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(
          getGPR(destReg) | getGPR(srcReg),
          getGPR(destReg)
        )
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.XOR_REG_WITH_REG: {
        const destReg = checkGPR(getNextOpcode())
        const srcReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(
          getGPR(destReg) ^ getGPR(srcReg),
          getGPR(destReg)
        )
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.NOT_REG: {
        const destReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(~getGPR(destReg), getGPR(destReg))
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.ROL_REG: {
        const destReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(
          exp<number>(() => {
            const value = getGPR(destReg)
            const MSB = Math.floor(value / 0x80)
            return (value << 1) + MSB
          }),
          getGPR(destReg)
        )
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.ROR_REG: {
        const destReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(
          exp<number>(() => {
            const value = getGPR(destReg)
            const LSB = value % 2
            return LSB * 0x80 + (value >> 1)
          }),
          getGPR(destReg)
        )
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.SHL_REG: {
        const destReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(getGPR(destReg) << 1, getGPR(destReg))
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
      case Opcode.SHR_REG: {
        const destReg = checkGPR(getNextOpcode())
        const [result, flags] = checkOperationResult(getGPR(destReg) >> 1, getGPR(destReg))
        setGPR(destReg, result)
        setSR(flags)
        incIP()
        break
      }
    }
  })
