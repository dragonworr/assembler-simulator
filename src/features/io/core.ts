export const SKIP = -1

export const MAX_PORT = 0x0f

export enum InputPort {
  SimulatedKeyboard = 0,
  Thermostat = 3,
  Keyboard = 7,
  NumericKeypad = 8
}

export type InputData =
  | {
      content: number
      port: InputPort
    }
  | {
      content: null
      port: null
    }

export interface InputSignals {
  data: InputData
  interrupt: boolean
  // TODO: interruptVectorAddress
}

export const initialInputSignals: InputSignals = {
  data: {
    content: null,
    port: null
  },
  interrupt: false
}

export enum OutputPort {
  TrafficLights = 1,
  SevenSegmentDisplay = 2,
  Heater = 3,
  SnakeInMaze = 4,
  StepperMotor = 5,
  Lift = 6,
  Keyboard = 7,
  NumericKeypad = 8
}

interface OutputData {
  content: number
  port: OutputPort
}

export type OutputSignals = Partial<{
  halted: true
  requiredInputPort: InputPort
  data: OutputData
  closeWindows: true
}>

export interface Signals {
  input: InputSignals
  output: OutputSignals
}
