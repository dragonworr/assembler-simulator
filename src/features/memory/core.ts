import type { AddressToMachineCodeMap, AddressToStatementMap } from '../assembler/core'
import { Mnemonic } from '../../common/constants'

export type MemoryData = number[]

export const VDU_START_ADDRESS = 0xc0

export const initData = (): MemoryData =>
  Array.from({ length: 0x100 }, (_, address) => (address < VDU_START_ADDRESS ? 0 : 0x20))

export const initDataFrom = (map: AddressToMachineCodeMap): MemoryData => {
  const data = initData()
  for (const address in map) {
    data[address] = map[address]
  }
  return data
}

export const getSourceFrom = (map: AddressToStatementMap): string[] => {
  const source: string[] = Array.from({ length: 0x100 }, (_, address) =>
    address < VDU_START_ADDRESS ? Mnemonic.END : ''
  )
  for (const address in map) {
    const addressNumber = Number.parseInt(address)
    const statement = map[address]
    const { instruction, operands } = statement
    if (instruction.mnemonic === Mnemonic.DB) {
      const operand = operands[0]
      // if (operand.type === OperandType.Number)
      if (typeof operand.value === 'number') {
        source[address] = operand.rawValue
      } else {
        // if (operand.type === OperandType.String)
        operand.rawValue.split('').forEach((char, index) => {
          source[addressNumber + index] = char
        })
      }
    } else {
      source[address] = instruction.mnemonic
      const nextAddress = addressNumber + 1
      operands.forEach((operand, index) => {
        const { rawValue } = operand
        // Address or RegisterAddress
        source[nextAddress + index] = operand.type.endsWith('Address') ? `[${rawValue}]` : rawValue
      })
    }
  }
  return source
}
