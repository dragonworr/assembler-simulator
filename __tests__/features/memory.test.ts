import { memorySerializer } from '../snapshotSerializers'
import { assemble } from '../../src/features/assembler/core'
import { getSourceFrom } from '../../src/features/memory/core'

expect.addSnapshotSerializer(memorySerializer)

describe('getSourceFrom', () => {
  it('should return correct source', () => {
    const [, addressToStatementMap] = assemble('jmp start start: add al, 01 mov bl, [cl] end')
    const source = getSourceFrom(addressToStatementMap)
    expect(source).toMatchSnapshot()
  })

  describe('with DB', () => {
    it('should display number', () => {
      const [, addressToStatementMap] = assemble('db a0 end')
      const source = getSourceFrom(addressToStatementMap)
      expect(source).toMatchSnapshot()
    })

    it('should split string', () => {
      const [, addressToStatementMap] = assemble('db "Hello, world!" end')
      const source = getSourceFrom(addressToStatementMap)
      expect(source).toMatchSnapshot()
    })
  })
})
