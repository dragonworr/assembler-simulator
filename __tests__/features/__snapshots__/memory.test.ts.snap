// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSourceFrom should return correct source 1`] = `
Array [
  JMP, START, ADD, AL, 01, MOV, BL, [CL], END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  , , , , , , , , , , , , , , , ,
  , , , , , , , , , , , , , , , ,
  , , , , , , , , , , , , , , , ,
  , , , , , , , , , , , , , , , ,
]
`;

exports[`getSourceFrom with DB should display number 1`] = `
Array [
  A0, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  , , , , , , , , , , , , , , , ,
  , , , , , , , , , , , , , , , ,
  , , , , , , , , , , , , , , , ,
  , , , , , , , , , , , , , , , ,
]
`;

exports[`getSourceFrom with DB should split string 1`] = `
Array [
  H, e, l, l, o, ,,  , w, o, r, l, d, !, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  , , , , , , , , , , , , , , , ,
  , , , , , , , , , , , , , , , ,
  , , , , , , , , , , , , , , , ,
  , , , , , , , , , , , , , , , ,
]
`;
